src/dsl.rs

```diff
@@ -15,6 +20,8 @@ pub struct Production<D: Domain> {
     pub arity: usize,
     pub lazy_args: HashSet<usize>,
     pub fn_ptr: Option<DSLFn<D>>,
```

Comment:
(do this later) have an enum containing the two kinds of pointers for functions. so like, fn_ptr and py_fn should just be one field fn_ptr.